src/corelibs/U2View/src/ov_sequence/sequence_info/SequenceInfo.cpp

```diff
@@ -353,30 +362,71 @@ void SequenceInfo::updateCodonsOccurrenceData(const QList<CharOccurResult> &codo
                                               : (isSequenceSelection ? tr("1 direct and 1 complement")
                                                                      : tr("3 direct and 3 complement"));
 
-    QString html;
-    html += "<table cellspacing=5>";
-    html += "<tr><td>" + tr("Report for:") + "</td><td><i>" + whereInfo + "</i></td></tr>";
-    html += "<tr><td>" + tr("Frames in report:") + "</td><td><i>" + frameInfo + "</i></td></tr>";
-    html += "</table>";
+    QString reportHeaderHtml;
+    reportHeaderHtml += "<table cellspacing=5>";
+    reportHeaderHtml += "<tr><td>" + tr("Report for:") + "</td><td><i>" + whereInfo + "</i></td></tr>";
+    reportHeaderHtml += "<tr><td>" + tr("Frames in report:") + "</td><td><i>" + frameInfo + "</i></td></tr>";
+    reportHeaderHtml += "</table>";
 
-    html += "<table cellspacing=5>";
-    if (codonStatList.isEmpty()) {
-        html += "<tr><td>" + tr("Selection is too small") + "</td></tr>";
+    QString codonsHtml = reportHeaderHtml;
+    QString aminoAcidsHtml = reportHeaderHtml;
+
+    if (codonStatMap.isEmpty()) {
+        QString selectionIsTooSmallHtml = "<tr><td>" + tr("Selection is too small") + "</td></tr>";
+        codonsHtml += selectionIsTooSmallHtml;
+        aminoAcidsHtml += selectionIsTooSmallHtml;
     } else {
         bool isFinal = codonTaskRunner.isIdle();
-        for (const CharOccurResult &codon : qAsConst(codonStatList)) {
-            html += "<tr>";
-            html += QString("<td><b>") + codon.getChar() + QString(":&nbsp;&nbsp;</b></td>");
-            html += "<td>" + getValue(getFormattedLongNumber(codon.getNumberOfOccur()), isFinal) + "&nbsp;&nbsp;</td>";
-            html += "</tr>";
+        // Prepare codons report & convert raw codons into amino acids.
+        QList<QByteArray> codons = codonStatMap.keys();
+        QHash<char, qint64> countPerAminoAcid;
+        bool tableColumnIndex = 0;    // Report is written using 2 columns.
+        for (const QByteArray &codon : qAsConst(codons)) {
+            codonsHtml += tableColumnIndex == 0 ? "<tr>" : "";
+            codonsHtml += QString("<td><b>") + QString::fromLatin1(codon) + QString(":&nbsp;&nbsp;</b></td>");
+            qint64 count = codonStatMap.value(codon);
+            codonsHtml += "<td>" + getValue(getFormattedLongNumber(count), isFinal) + "&nbsp;&nbsp;</td>";
+            codonsHtml += tableColumnIndex == 1 ? "</tr>" : "";
+            tableColumnIndex = (tableColumnIndex + 1) % 2;
+            if (aminoTranslation != nullptr) {
+                char acidChar = 0;
+                aminoTranslation->translate(codon.constData(), 3, &acidChar, 1);
+                if (acidChar != 0) {
+                    countPerAminoAcid[acidChar] += count;
+                }
+            }
+        }
+
+        // Prepare amino acids report.
+        QList<CharOccurResult> aminoAcidList;
+        for (auto it = countPerAminoAcid.begin(); it != countPerAminoAcid.end(); it++) {
```

Comment:
I prefer we never use iterators, cuz since C++11 we have range-based loops, which are easier to understand